proxy/coco/src/request.rs

```diff
@@ -0,0 +1,240 @@
+use std::collections::HashSet;
+use std::convert::TryFrom;
+use std::marker::PhantomData;
+use std::time::Duration;
+
+use librad::net::peer::types::Gossip;
+use librad::uri::RadUrn;
+
+pub mod waiting_room;
+
+type PeerId = String;
+
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub struct Repo {
+    urn: RadUrn,
+    name: String,
+}
+
+const MAX_QUERIES: usize = 1;
+const MAX_CLONES: usize = 1;
+const PERIOD: Duration = Duration::from_secs(1); // Not for the whole request but for re-request
+
+pub fn exponential_backoff(attempts: Attempts, interval: Duration) -> Duration {
+    let exp = u32::try_from(attempts.queries + attempts.clones).unwrap_or(u32::MAX);
+    Duration::from_millis(u64::pow(2, exp)) + interval
+}
+
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub enum State {
+    Found(PeerId),
+    Cloning(PeerId),
+    Cloned(Repo),
+}
+
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub struct Created;
+
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub struct Canceled;
+
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub struct TimedOut;
+
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub struct Requested;
+
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub struct Fulfilled;
+
+#[derive(Clone, Copy, Debug, Hash, PartialEq)]
+pub struct Attempts {
+    queries: usize, // how often we gossip
+    clones: usize,  // how often we try to clone
+}
+
+impl Attempts {
+    pub fn new() -> Self {
+        Attempts {
+            queries: 0,
+            clones: 0,
+        }
+    }
+}
+
+pub enum Attempt {
+    Query,
+    CloneRepo,
+}
+
+#[derive(Clone, Debug, thiserror::Error, PartialEq)]
+pub enum Error {
+    #[error("the URN found '{actual}' was not the expected URN '{expected}'")]
+    UrnMismatch { expected: RadUrn, actual: RadUrn },
+    #[error("the request was not found to be in the cloning state and so it cannot be fulfilled")]
+    ExpectedCloningState,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Request<S, T> {
+    urn: RadUrn,
+    state: PhantomData<S>,
+    found_peers: HashSet<PeerId>,
+    attempts: Attempts,
+    timestamp: T,
+    internal: Option<State>,
+}
+
+impl<S, T> From<Request<S, T>> for Gossip {
+    fn from(request: Request<S, T>) -> Self {
+        Self {
+            urn: request.urn,
+            rev: None,
+            origin: None,
+        }
+    }
+}
+
+impl<S, T> Request<S, T> {
+    fn coerce<R>(self, timestamp: T) -> Request<R, T> {
+        Request {
+            urn: self.urn,
+            state: PhantomData,
+            found_peers: self.found_peers,
```

Comment:
This answers my question above, we should start think about how we track which peer we cloned from as well and mark them when cloning failed.